src/include/miopen/graphapi/opgraph.h

```diff
@@ -0,0 +1,223 @@
+    {
+        assert(src);
+        if(!hasInEdge(src, tens_ptr))
+        {
+            mInEdges.emplace_back(src, tens_ptr);
+        }
+    }
+};
+
+using OpEdge = OpNode::Edge;
+
+class SourceOpNode : public OpNode
+{
+protected:
+    std::vector<Tensor*> mOutTensors;
+    friend class OpGraph;
+
+    const std::string& signName() const final
+    {
+        static const std::string s = "INTERNAL::SRC";
+        return s;
+    }
+
+    std::vector<Tensor*> getInTensors() const final { return {}; }
+
+    std::vector<Tensor*> getOutTensors() const final { return mOutTensors; }
+
+    bool hasOutTensor(Tensor* tensor) const { return internal::contains(mOutTensors, tensor); }
+
+    void addOutTensor(Tensor* tens_ptr)
+    {
+        assert(!hasOutTensor(tens_ptr));
+        mOutTensors.emplace_back(tens_ptr);
+    }
+};
+
+class SinkOpNode : public OpNode
+{
+protected:
+    std::vector<Tensor*> mInTensors;
+    friend class OpGraph;
+
+    const std::string& signName() const final
+    {
+        static const std::string s = "INTERNAL::SINK";
+        return s;
+    }
+
+    std::vector<Tensor*> getInTensors() const final { return mInTensors; }
+
+    std::vector<Tensor*> getOutTensors() const final { return {}; }
+
+    bool hasInTensor(Tensor* tensor) const { return internal::contains(mInTensors, tensor); }
+
+    void addInTensor(Tensor* tens_ptr)
+    {
+        assert(!hasInTensor(tens_ptr));
+        mInTensors.emplace_back(tens_ptr);
+    }
+};
+
+class OpGraph
+{
+    SourceOpNode mSrcNode{};
+    SinkOpNode mSinkNode{};
+    std::vector<OpNode*> mNodes{};
+
+public:
+    bool hasNode(OpNode* n) const { return internal::contains(mNodes, n); }
+
+    bool hasEdge(OpNode* src, Tensor* tens_ptr, OpNode* dst) const
+    {
+        assert(src);
+        assert(dst);
+        return src->hasOutEdge(dst, tens_ptr) && dst->hasInEdge(src, tens_ptr);
+    }
+
+private:
+    friend class OpGraphBuilder;
+
+    void addNodes(std::vector<OpNode*>&& nodes) { mNodes = std::move(nodes); }
+
+    void addEdge(OpNode* src, Tensor* tens_ptr, OpNode* dst)
+    {
+        assert(src);
+        assert(dst);
+        src->addOutEdge(dst, tens_ptr);
+        dst->addInEdge(src, tens_ptr);
+    }
+
+    void addEdgeFromSrc(OpNode* dst, Tensor* tens_ptr)
+    {
+        mSrcNode.addOutTensor(tens_ptr);
+        addEdge(&mSrcNode, tens_ptr, dst);
+    }
+
+    void addEdgeToSink(OpNode* src, Tensor* tens_ptr)
+    {
+        mSinkNode.addInTensor(tens_ptr);
+        addEdge(src, tens_ptr, &mSinkNode);
+    }
+};
+
+class OpGraphBuilder
+{
+private:
+    std::vector<OpNode*> mNodes;
+
+public:
+    bool hasNode(OpNode* node) const { return internal::contains(mNodes, node); }
+
+    void addNode(OpNode* node)
+    {
+        assert(!hasNode(node));
+        mNodes.emplace_back(node);
+    }
+
+    struct EdgeInfo
+    {
+        OpNode* mSrc = nullptr;
+        std::vector<OpNode*> mDests{};
+    };
+
+    OpGraph build();
```

Comment:
You can't call this method multiple times and it's not obvious.
Either it should be obvious and bullet proof, or `mNodes` must not be moved.

Reply:
OK, so I forgot to make it an `r-value` method. That's the general pattern we will have with `builders` that `build()` will be an r-value method that consumes the calling builder object.